operator/rediscontroller/controller_test.go

```diff
@@ -0,0 +1,339 @@
+// Therefore we just reset the ExpectedCalls (without mutex since that is private) and re-add the mocked calls
+// which are always necessary to have.
+func (ts *RedisControllerTestSuite) resetExoClientMock() {
+	ts.ExoClientMock.ExpectedCalls = nil
+	ts.mockCaCertificateResponse()
+}
+
+func (ts *RedisControllerTestSuite) mockCaCertificateResponse() *mock.Call {
+	return ts.ExoClientMock.On("GetDbaasCaCertificateWithResponse", mock.Anything).
+		Return(oapi.ParseGetDbaasCaCertificateResponse(&http.Response{
+			StatusCode: http.StatusOK,
+			Header:     http.Header{"Content-Type": []string{"application/json"}},
+			Body:       io.NopCloser(strings.NewReader(`{"certificate": "certifyingliketheresnotomorrow"}`)),
+		}))
+}
+
+func (ts *RedisControllerTestSuite) getRedisResponse(mg *exoscalev1.Redis, settings map[string]any) *oapi.GetDbaasServiceRedisResponse {
+	providerSpec := mg.Spec.ForProvider
+	master := oapi.DbaasNodeStateRoleMaster
+	running := oapi.EnumServiceStateRunning
+
+	return &oapi.GetDbaasServiceRedisResponse{
+		JSON200: &oapi.DbaasServiceRedis{
+			Maintenance: &oapi.DbaasServiceMaintenance{
+				Dow:  providerSpec.Maintenance.DayOfWeek,
+				Time: providerSpec.Maintenance.TimeOfDay.String(),
+			},
+			Name: oapi.DbaasServiceName(mg.Name),
+			NodeStates: &[]oapi.DbaasNodeState{
+				{
+					Name:  "test-1",
+					Role:  &master,
+					State: oapi.DbaasNodeStateStateRunning,
+				},
+			},
+			IpFilter:              &[]string{providerSpec.IPFilter[0]},
+			Plan:                  providerSpec.Size.Plan,
+			RedisSettings:         &settings,
+			State:                 &running,
+			TerminationProtection: pointer.Bool(false),
+			ConnectionInfo: &struct {
+				Password *string   `json:"password,omitempty"`
+				Slave    *[]string `json:"slave,omitempty"`
+				Uri      *[]string `json:"uri,omitempty"`
+			}{
+				Uri: &[]string{"https://foo:bar@baz:5321"},
+			},
+			Uri:       nil,
+			UriParams: nil,
+			Users:     nil,
+			Version:   nil,
+		},
+	}
+}
+
+func newRedisInstance(name string, settings map[string]any) *exoscalev1.Redis {
+	rs, err := mapper.ToRawExtension(&settings)
+	if err != nil {
+		panic(fmt.Errorf("settings not convertible to raw extension: %w", err))
+	}
+
+	return &exoscalev1.Redis{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: exoscalev1.RedisSpec{
+			ResourceSpec: xpv1.ResourceSpec{
+				WriteConnectionSecretToReference: &xpv1.SecretReference{
+					Name:      name,
+					Namespace: "default",
+				},
+			},
+			ForProvider: exoscalev1.RedisParameters{
+				Maintenance: exoscalev1.MaintenanceSpec{
+					DayOfWeek: oapi.DbaasServiceMaintenanceDowMonday,
+					TimeOfDay: "12:00:00",
+				},
+				Zone: "ch-dk-2",
+				DBaaSParameters: exoscalev1.DBaaSParameters{
+					TerminationProtection: false,
+					Size: exoscalev1.SizeSpec{
+						Plan: "testplan",
+					},
+					IPFilter: exoscalev1.IPFilter{"0.0.0.0/0"},
+				},
+				RedisSettings: rs,
+			},
+		},
+	}
```

Comment:
Something that I started doing recently with Kubernetes objects. Instead of wasting tons of lines with spec and braces, I do this instead:

```suggestion
	i := &exoscalev1.Redis{}
	i.Name = name
	i.Spec.ForProvider.Maintenance.DayOfWeek = oapi.DbaasServiceMaintenanceDowMonday
	i.Spec.ForProvider.Maintenance.TimeOfDay = "12:00:00"
	...
	return i
```
it works well for structs that aren't pointers.

No hard feeling though

Reply:
I don't think that's an improvement TBH